clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/CallbackMap.java

```diff
@@ -3,30 +3,18 @@
 
 package com.microsoft.aspnet.signalr;
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.List;
+import java.util.*;
 import java.util.concurrent.ConcurrentHashMap;
 
 class CallbackMap {
-    private ConcurrentHashMap<String, List<InvocationHandler>> handlers = new ConcurrentHashMap<>();
+    private Map<String, List<InvocationHandler>> handlers = new ConcurrentHashMap<>();
 
-    public InvocationHandler put(String target, ActionBase action, ArrayList<Class<?>> classes) {
-        InvocationHandler handler = new InvocationHandler(action, Collections.unmodifiableList(classes));
-
-        handlers.computeIfPresent(target, (methodName, handlerList) -> {
-            handlerList.add(handler);
-            return handlerList;
-        });
-        handlers.computeIfAbsent(target, (ac) -> new ArrayList<>(Arrays.asList(handler)));
+    public InvocationHandler put(String target, ActionBase action, Class<?>... classes) {
+        InvocationHandler handler = new InvocationHandler(action, classes);
+        handlers.computeIfAbsent(target, ac -> new ArrayList<>()).add(handler);
```

Comment:
I don't think this is thread safe. You're calling add on a List

Reply:
Do we support calling `.on` from multiple threads?

Reply:
Technically yes, since we allow you to add or remove handlers at any time

Reply:
I think we need to use [`compute`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#compute-K-java.util.function.BiFunction-) here. The `BiFunction` taken by compute gets the current value or `null` if there is no value, and it guarantees that the entire method is invoked atomically. So the code would be something like:

```java
handlers.compute((key, value) -> {
    if (value == null) {
        value = new ArrayList<>();
    }
    value.add(handler);
    return value;
});
```

That way we, in a single atomic invocation, initialize the list if there isn't one already and then add the item to the list. Since `compute` guarantees atomicity, other attempts to add to the list will be blocked while we initialize the list and then will see the updated list. At least that's how I understand the Javadocs.